File: typical90/rust/src/bin/007.rs
use proconio::{fastout, input};
use superslice::*;

#[fastout]
fn main() {
    input! {
        n: usize,
        mut a: [i32; n],
        q: usize,
        b: [i32; q]
    }
    // u32だとl31でパニックする
    // println!("{:?}, {:?}, {:?}, {:?}", b, a, q, b)
    // 愚直にやったらO(nq)=10^10
    // rustのsortはmerge sortらしい
    a.sort();
    for b_i in b {
        // supersliceでlower_boundを使えるようになる
        // &b_i以上の要素のなかで最小の要素のindex
        let i = a.lower_bound(&b_i);
        // i = 0 のときはaの全要素がb_iよりおおきいのでいちばん近い要素の上限はa[0]
        if i == 0 {
            let a1 = (a[0] - b_i).abs();
            println!("{}", a1);
            continue;
        }
        // i >= n のときはaの全要素がb_iよりちいさいので上限はなくて一番近い要素はa[n-1]
        if i >= n {
            let a2 = (a[n - 1] - b_i).abs();
            println!("{}", a2);
            continue;
        }
        let a1 = (a[i] - b_i).abs();
        let a2 = (a[i - 1] - b_i).abs();
        let ans = a1.min(a2);
        println!("{}", ans)
    }
}

File: typical90/python/007.py
import bisect

# 全体ではO((Q+N)logN)

# A_i (1 <= i <= N)
# B_j (1 <= j <= Q)
# j = 1, 2, 3, ..., Qそれぞれについて番号jの生徒の不満度として考えられる最小値
N = int(input())
A = list(map(int, input().split()))
Q = int(input())
B = [int(input()) for _ in range(Q)]


# O(NlogN)
A.sort()

# O((Qlog N)
for b_i in B:
    i = bisect.bisect_left(A, b_i)
    if i == 0:
        a1 = abs(A[0] - b_i)
        print(a1)
        continue

    if i >= N:
        a2 = abs(A[N - 1] - b_i)
        print(a2)
        continue

    a1 = abs(A[i] - b_i)
    a2 = abs(A[i - 1] - b_i)
    ans = min(a1, a2)
    print(ans)
